src/persist-client/src/internal/compact.rs

```diff
@@ -458,6 +430,218 @@ where
+                metrics.compaction.applied.inc();
+                metrics.compaction.applied_subset_match.inc();
+                machine.applier.shard_metrics.compaction_applied.inc();
+            }
+            ApplyMergeResult::NotAppliedNoMatch
+            | ApplyMergeResult::NotAppliedInvalidSince
+            | ApplyMergeResult::NotAppliedTooManyUpdates => {
+                if let ApplyMergeResult::NotAppliedTooManyUpdates = &apply_merge_result {
+                    metrics.compaction.not_applied_too_many_updates.inc();
+                }
+                metrics.compaction.noop.inc();
+                let mut part_deletes = PartDeletes::default();
+                for part in &res.output.parts {
+                    part_deletes.add(part);
+                }
+                part_deletes
+                    .delete(
+                        machine.applier.state_versions.blob.as_ref(),
+                        machine.shard_id(),
+                        GC_BLOB_DELETE_CONCURRENCY_LIMIT.get(&machine.applier.cfg),
+                        &*metrics,
+                        &metrics.retries.external.compaction_noop_delete,
+                    )
+                    .await;
+            }
+        };
+
+        Ok(maintenance)
+    }
+
+    /// Compacts input batches in bounded memory.
+    ///
+    /// The memory bound is broken into pieces:
+    ///     1. in-progress work
+    ///     2. fetching parts from runs
+    ///     3. additional in-flight requests to Blob
+    ///
+    /// 1. In-progress work is bounded by 2 * [BatchBuilderConfig::blob_target_size]. This
+    ///    usage is met at two mutually exclusive moments:
+    ///   * When reading in a part, we hold the columnar format in memory while writing its
+    ///     contents into a heap.
+    ///   * When writing a part, we hold a temporary updates buffer while encoding/writing
+    ///     it into a columnar format for Blob.
+    ///
+    /// 2. When compacting runs, only 1 part from each one is held in memory at a time.
+    ///    Compaction will determine an appropriate number of runs to compact together
+    ///    given the memory bound and accounting for the reservation in (1). A minimum
+    ///    of 2 * [BatchBuilderConfig::blob_target_size] of memory is expected, to be
+    ///    able to at least have the capacity to compact two runs together at a time,
+    ///    and more runs will be compacted together if more memory is available.
+    ///
+    /// 3. If there is excess memory after accounting for (1) and (2), we increase the
+    ///    number of outstanding parts we can keep in-flight to Blob.
+    pub fn compact_stream(
+        cfg: CompactConfig,
+        blob: Arc<dyn Blob>,
+        metrics: Arc<Metrics>,
+        shard_metrics: Arc<ShardMetrics>,
+        isolated_runtime: Arc<IsolatedRuntime>,
+        req: CompactReq<T>,
+        write_schemas: Schemas<K, V>,
+    ) -> impl Stream<Item = Result<CompactRes<T>, anyhow::Error>> {
```

Comment:
Let's have this return `HollowBatch` instead of `CompactRes` for now, since it is not actually safe to apply to the spine.